[2주차] 리팩토링 및 검색 기능 추가

kdmstj opened the pull request:
## [2주차] 리팩토링 및 검색 기능 추가

### 작업 내용  
- 멀티 모듈을 사용하여 헥사고날 아키텍처 수정
   - `api` 모듈에 있었던 ApiApplication 을 웹 진입점인 `adapter.web` 모듈에 있는 것이 맞다고 판단하여 옮겨주었습니다.
   - 역할이 모호했던 `api` 모듈을 삭제해주었습니다.
   - 최종적으로 다음과 같은 아키텍처로 수정하였습니다. 
- 검색 기능과 필터링 기능을 추가해주었습니다.
   - @RequestParam 어노테이션을 사용하여 query parameter 로 keyword 및 genre 를 요청받도록 하였습니다.
   - @Size 어노테이션을 사용하여 요청된 값에 대해 유효성 판단을 하도록 하였습니다.
   - 동적으로 쿼리문을 작성하기 위해서 QueryDSL 을 적용해주었습니다.
       - DB 에서 쿼리문을 통해 조인된 결과 값을 받고, 이를 CustomImpl에서 가공해서 원하는 응답값으로 응답할 수 있도록 하였습니다. 
- movie, show_time 테이블에 인덱스를 추가해주었습니다. 
    - movie 테이블 : idx_movie_genre_release_at_title 인덱스 생성
    - show_time 테이블 : idx_movie_genre_release_at_title 인덱스 생성   
    - 인덱스를 적용하고 title 검색 시 LIKE 연산과 동등 연산자를 나눠서 성능 테스트를 진행하였습니다. 
- 성능 테스트 결과 보고서는 [이곳](https://www.notion.so/17fe5ab6924e80109e6efc1149612b21?pvs=4)에서 확인할 수 있습니다. 

### 발생했던 문제와 해결 과정을 남겨 주세요. 
- **문제** : EntityScan 시 Entity 를 찾지 못하는 에러 발생하였습니다. 
- **해결 방법** : 원인은 javax 라이브러리를 import 하였기 때문이었습니다. 이를 해결하기 위해서 jakarta 라이브러리를 import 해주었습니다.

### 이번 주차에서 고민되었던 지점이나, 어려웠던 점을 알려 주세요.
-  개봉 중인 영화에 대하여 show_time 을 left join 시 (영화1, show_time1), (영화1, show_time2) 으로 결과가 반환하여 직접 영화에 대해 grouping 을 하여 응답값에 담아주도록 하였는데 괜찮은 방법인지 궁금합니다. 
-  index 를 설정할 때 성능 테스트를 통해 여러 경우를 고민했습니다.
    - 단일 index 를 여러개를 할 것인지 복합 index 를 할 것인지에 대한 고민
    - 어떤 컬럼을 index 로 설정할지에 대한 고민 
    - 복합 index 를 설정할 때는 어떤 순서대로 하면 좋을지에 대한 고민

### 리뷰 포인트
- 저에게는 다소 생소하고 어려운 부분이 있어서 진도가 늦어지고 있습니다.. 그럼에도 불구하고 열심히 저의 속도로 따라가려고 노력하고 있습니다! 진행된 부분까지 리뷰해주셔서 감사합니다 :)

### 기타 질문

kdmstj committed:
fix: JavaEE에서 Jakarta EE로 전환

- EntityScan 시 Entity 를 찾지 못하는 에러 발생하였으며, 원인은 javax 라이브러리 @Entity 를 import 하였기 때문이었습니다. 이를 해결하기 위해서 jakarta 라이브러리를 import 해주었습니다.

kdmstj committed:
fix: ApiApplication 모듈 위치 변경

- 웹 진입접인 adapter.api 로 이동해주었습니다.

kdmstj committed:
fix: ApiApplication 모듈 위치 변경

- 웹 진입접인 adapter.api 로 이동해주었습니다.

kdmstj committed:
chore: adpater builde.gradle 삭제 및 변경

- adapter 모듈 내 web, persistence 모듈이 있으므로 해당 디렉토리의 build.gradle 은 삭제해주었습니다.

kdmstj committed:
fix: Mapper Component 어노테이션 추가

kdmstj committed:
Merge branch 'week1' of https://github.com/kdmstj/redis_1st into week1

kdmstj committed:
chore: settings.gradle api 모듈 삭제

- api 모듈이 삭제됨에 따라 settings.gradle 에서도 삭제 처리해 주었습니다.

kdmstj committed:
Update README.md

kdmstj committed:
fix: db table 명 변경

kdmstj committed:
fix: BaseEntity 생성일, 수정일 필드명 변경

kdmstj committed:
chore: mysql timezone 설정

- mysql timezone 을 seoul 로 설정하였습니다.

kdmstj committed:
add: 응답값에 movie showtime 추가

- movie 와 left join 하여 showtime 과 그에 해당하는 theater 를 조회하여 배열에 담았습니다.
- service 계층에서 response 를 만들도록 하였습니다.

kdmstj committed:
add: 검색/필터 기능 추가

- 동적으로 keyword 와 genre 를 받기 위해서 querydsl 을 적용하였습니다.
- keyword 의 유효성 검사를 위해서 validation 라이브러리를 추가해주었습니다.

kdmstj committed:
chore: 응답값 한글 인코딩

kdmstj committed:
chore: http 요청

- movie 조회
- filter 조회
- keyword 조회

soonhankwon commented:
안녕하세요 은서님 😃
2주차 열심히 참여해주셔서 감사합니다. 리뷰 진행하겠습니다 :)

soonhankwon commented on cinema/adapter/persistence/src/main/java/com/hanghae/cinema/adapter/persistence/MoviePersistenceAdapter.java:
```diff
@@ -1,27 +1,22 @@
-    private final MovieEntityMapper movieEntityMapper;
 
-    @Override
-    public List<Movie> getMovieList(LocalDate date) {
```
메서드 시그네처의 반환타입이 List인데 메서드 네임에 또 List가 들어가는것은 불필요하다고 생각하는편입니다 :)

kdmstj replied:
메서드 이름에 대해서 한번 더 공부해보겠습니다 :)
말씀해주신 것처럼 반환타입, 파라미터에 명시된 정보는 메서드 네임에 중복되지 않도록 수정하겠습니다!

soonhankwon commented on cinema/adapter/persistence/src/main/java/com/hanghae/cinema/adapter/persistence/entity/MovieEntity.java:
```diff
@@ -29,7 +29,7 @@ public class MovieEntity extends BaseEntity {
     private LocalDate releaseAt;
 
-    private Integer runningTime;
+    private Integer runningTime; //시간 단위 (분)
```
주석은 잘 붙여주셨지만, 시간 단위가 분이라는 것을 변수만  보고 알기 어렵습니다 :)

kdmstj replied:
시간 단위가 분이라는 것을 변수만 보고 알 수 있도록 변수명에 시간 단위가 나타나도록 수정하도록 하겠습니다!

soonhankwon commented on cinema/adapter/persistence/src/main/java/com/hanghae/cinema/adapter/persistence/repository/MovieRepository.java:
```diff
@@ -1,14 +1,28 @@
+                            where m.releaseAt <= :date and s.startedAt > :dateTime
+                            order by m.releaseAt DESC, s.startedAt ASC
+            """)
+    List<Object[]> findMoviesWithShowTimesAndTheaters(LocalDate date, LocalDateTime dateTime);
```
Object를 반환타입으로 사용하면 런타임에 예외가 발생할 가능성이 매우 커집니다.
- QueryDSL을 사용하고 계신데 DTO를 활용한 Projection을 사용하는것이 어떨까요?

kdmstj replied:
Object 을 반환 타입으로 사용하면 예외 발생할 가능성이 높아 지양해야 하군요!
QueryDSL 을 사용하기 전의 코드인데 미처 지우지 못했습니다! 

추가적으로 QueryDSL 에서 DTO 를 통해서 응답하도록 하였는데, 조회 시 (영화1, 시간1), (영화1, 시간2) 이런식으로 조회가 되어 직접 Map 으로 가공하여서 (영화1, (시간1, 시간2))로 응답하도록 하였습니다. 해당 방법이 괜찮을까요? 혹시 좋은 방법이 있을까요!?

soonhankwon commented on cinema/adapter/persistence/src/main/java/com/hanghae/cinema/adapter/persistence/repository/MovieRepository.java:
```diff
@@ -1,14 +1,28 @@
+import java.time.LocalDateTime;
 import java.util.List;
 
 @Repository
```
Spring data jpa를 사용하고 있는데 Repository 애노테이션은 불필요하다고 생각합니다.

kdmstj replied:
넵 수정하도록 하겠습니다!

soonhankwon commented on cinema/adapter/persistence/src/main/java/com/hanghae/cinema/adapter/persistence/repository/MovieRepositoryCustomImpl.java:
```diff
@@ -0,0 +1,99 @@
+                        theaterEntity.name
+                )
+                .from(movieEntity)
+                .leftJoin(showTimeEntity).on(movieEntity.id.eq(showTimeEntity.movie.id))
```
상영정보 없음을 고려하신 leftjoin인가요? :)

kdmstj replied:
넵 맞습니다. 현재 존재하고 개봉된 영화에 대해서만 show_time 을 조회할 수 있도록 하였습니다:)

soonhankwon commented on cinema/adapter/web/src/main/java/com/hanghae/cinema/adapter/web/controller/MovieController.java:
```diff
@@ -19,7 +22,14 @@ public class MovieController {
+            @Size(max = 255, message = "영화 제목은 최대 255자 이하여야 합니다.")
+            String keyword,
+
+            @RequestParam(required = false)
```
Genre에 장르만 정의되어 있는데, 전체검색은 못하는건지? 의도하신 부분인지 궁금하네요 :)

kdmstj replied:
질문하신 내용이 장르를 String 으로 검색할 수 없도록 의도한건지에 대한 것이 맞을까요~?

현재 MovieGenre 는 Enum 으로 선언되어 있으며, 클라이언트가 Enum 에 선언된 value 값 하나를 요청해서 필터링한다는 점에서 MovieGenre 로 받도록 하였습니다!

soonhankwon replied:
required가 false였었군요 좋습니다 :)

soonhankwon commented on cinema/application/src/main/java/com/hanghae/cinema/application/dto/MovieResponseDTO.java:
```diff
@@ -2,11 +2,18 @@
+@Getter
+@AllArgsConstructor
+@NoArgsConstructor
 public class MovieResponseDTO {
```
필요한 필드들이 잘 정의되어 있어서 좋네요 👍

soonhankwon commented:
바쁘셔서 요구사항을 많이 구현 및 적용하시지 못하신듯해서 아쉽습니다 :)

### 좋았던 점

- 하나의 API를 통해 키워드, 장르 필터링을 할 수 있도록 설계해주신 점이 좋습니다.
- DockerFile에 Charset, Collation, Timezone 설정을 잘 해주셨습니다.

### 아쉬웠던 점

- 아직 인덱스, 캐싱에 대한 부분에 대한 2주차 요구사항이 적용 및 설명되어 있지 않아서 아쉽습니다.
    - 매우 중요한 부분이니 시간이 되시면 구현 및 테스트해보는 경험을 가지면 좋겠습니다.
- 메서드 시그네처가 개인적으로 생각하기에 다소 과한부분이 존재합니다.

kdmstj committed:
fix: movie 조회 메서드 네임 수정

- 반환타입으로 확인할 수 있는 것에 대해 메서드 네임에 중복해서 작성하는 것을 지양하도록 네임을 수정하였습니다.

kdmstj committed:
Update README.md

kdmstj committed:
fix: MovieRepository @Repository 어노테이션 삭제

- @EnableJpaRepository 에 의해서 JpaRepository 를 상속한 인터페이스는 Repository 에 등록되므로 불필요하다고 판단하여 삭제하였습니다.

kdmstj committed:
fix: Movie.runningTime 변수명 변경

- runningTime 의 시간 단위롤 변수명에 나타냄으로써 직관적으로 시간 단위를 확인할 수 있도록 하였습니다.

kdmstj committed:
fix: MovieRepository 사용하지 않는 메소드 삭제

- queryDSL 을 통한 조회로 사용하지 않은 메소드는 삭제하였습니다.

kdmstj committed:
add: 로컬 캐싱 적용

- caffeine keyword_genre 에 대해서 로컬 캐싱을 적용하여 성능을 높이도록 하였습니다.
